### PR TITLE
Add edge cases to initialization function checks

### PR DESCRIPTION
**Pull request by mariaschuld**
**Context:**

Many templates require a special shape of input parameters when run with 1 wire. For zero wires, all init functions should return an empty array. 

So far, the tests did not consistently check these edge cases. 

**Description of the Change:**

Rewrite ``test_init.py`` to test edge cases. For this the main fixture was rewritten to explicitly list all inputs.

The fix exposed 2 issues in the init functions which were fixed.

**Benefits:** 

Be sure that template init functions work for edge cases.

**Possible Drawbacks:** 

The fixture in the test is getting terribly long, but this will be addressed by changes in the template format soon, see issues #574 and #575.

**Related GitHub Issues:** #561

### CODE REVIEW

File: tests/test_init.py [hunk 1]
```diff
@@ -19,90 +19,241 @@
...
-                      (qml.init.simplified_two_design_initial_layer_normal, {'n_wires': 3, 'mean': 0, 'std': 1}, (3,)),
-                      (qml.init.simplified_two_design_weights_uniform, {'n_layers': 2, 'n_wires': 1, 'low': 0, 'high': 1}, (0,)),
-                      (qml.init.simplified_two_design_weights_uniform, {'n_layers': 2, 'n_wires': 2, 'low': 0, 'high': 1}, (2, 1, 2)),
-                      (qml.init.simplified_two_design_weights_uniform, {'n_layers': 2, 'n_wires': 4, 'low': 0, 'high': 1}, (2, 3, 2)),
-                      (qml.init.simplified_two_design_weights_normal, {'n_layers': 2, 'n_wires': 1, 'mean': 0, 'std': 1}, (0, )),
-                      (qml.init.simplified_two_design_weights_normal, {'n_layers': 2, 'n_wires': 2, 'mean': 0, 'std': 1}, (2, 1, 2)),
-                      (qml.init.simplified_two_design_weights_normal, {'n_layers': 2, 'n_wires': 4, 'mean': 0, 'std': 1}, (2, 3, 2)),
-                      (qml.init.basic_entangler_layers_normal, {'n_layers': 2, 'n_wires': 1, 'mean': 0, 'std': 1},
+INIT_KWARGS_SHAPES = [(qml.init.random_layers_normal,
```
Comment (mariaschuld) [inline tests/test_init.py:28]: There should now be ``wires=1`` cases for every init function.

File: pennylane/init.py [hunk 2]
```diff
@@ -161,9 +161,14 @@ def random_layers_uniform(n_layers, n_wires, n_rots=None, low=0, high=2 * pi, se
     if seed is not None:
         np.random.seed(seed)
 
+    # set default
     if n_rots is None:
         n_rots = n_wires
 
+    # no circuit if there are no wires
+    if n_wires == 0:
```
Comment (antalszava) [inline pennylane/init.py:169]: When did these cases arise? (perhaps I had this question before :smile:) Wondering if it's a good call to "silently" return an empty array as the output of `np.random.uniform`. Would a warning work here or that would be too much?

File: tests/test_init.py [hunk 3]
```diff
@@ -19,90 +19,241 @@
...
+                      (qml.init.random_layers_normal,
+                       {'n_layers': 2, 'n_wires': 1, 'n_rots': 10, 'mean': 0, 'std': 1},
+                       (2, 10)),
+                      (qml.init.random_layers_normal,
+                       {'n_layers': 2, 'n_wires': 3, 'n_rots': None, 'mean': 0, 'std': 1},
+                       (2, 3)),
+                      (qml.init.strong_ent_layers_normal,
+                       {'n_layers': 2, 'n_wires': 3, 'mean': 0, 'std': 1},
+                       (2, 3, 3)),
+                      (qml.init.strong_ent_layers_normal,
```
Comment (antalszava) [inline tests/test_init.py:40]: Out of scope, just for my own understanding: what is the difference between:
* `cvqnn_layers_theta_uniform`
* `cvqnn_layers_phi_uniform`
* `cvqnn_layers_varphi_uniform`?

Are they separated such upon seeding one gets different results?

File: tests/test_init.py (hunk 1 above)
Comment (antalszava) [inline tests/test_init.py:28]: (Mino suggestion): placing small comments in the list could help out with readability. If this would become irrelevant with the changes that are planned, then this is not crucial :)
```suggestion
INIT_KWARGS_SHAPES = [
                       # Random layers parameters
                       (qml.init.random_layers_normal,
```

File: tests/test_init.py [hunk 4]
```diff
@@ -19,90 +19,241 @@
...
+                       (3,)),
+                      (qml.init.interferometer_varphi_normal,
+                       {'n_wires': 1, 'mean': 0, 'std': 1},
+                       (1,)),
+                      (qml.init.random_layers_uniform,
+                       {'n_layers': 2, 'n_wires': 3, 'n_rots': 10, 'low': 0, 'high': 1},
+                       (2, 10)),
+                      (qml.init.random_layers_uniform,
+                       {'n_layers': 2, 'n_wires': 3, 'n_rots': None, 'low': 0, 'high': 1},
+                       (2, 3)),
```
Comment (antalszava) [inline tests/test_init.py:114]: Just to have a case of falling back to `n_wires=1` due to `n_rots is None`
```suggestion
                       (2, 3)),
                       (qml.init.random_layers_uniform,
                       {'n_layers': 2, 'n_wires': 1, 'n_rots': None, 'low': 0, 'high': 1},
                       (2, 1)),
```

File: tests/test_init.py (hunk 1 above)
Comment (antalszava) [reply tests/test_init.py:28]: My suggestion for this list in the short term would be to split it up based on the function that is being called:
```python
RANDOM_NORMAL = [
                       (qml.init.random_layers_normal,
                       {'n_layers': 2, 'n_wires': 3, 'n_rots': 10, 'mean': 0, 'std': 1},
                       (2, 10)),
                      (qml.init.random_layers_normal,
                       {'n_layers': 2, 'n_wires': 1, 'n_rots': 10, 'mean': 0, 'std': 1},
                       (2, 10)),
                      (qml.init.random_layers_normal,
                       {'n_layers': 2, 'n_wires': 3, 'n_rots': None, 'mean': 0, 'std': 1},
                       (2, 3))
                       ]

# All the other types defined in a similar fashion

# Grouping terms within the base group
INIT_KWARGS_SHAPES = [
                       *RANDOM_NORMAL, *RANDOM_UNIFORM,
                       *STRONG_ENT_NORMAL, *STRONG_ENT_UNIFORM,
                       (....)
                       ]
```
This will ensure that there is a clear logical separation between cases and improve readability. It does not tackle the length of these lists but makes it easier to have an overview.

File: tests/test_init.py [hunk 5]
```diff
@@ -19,90 +19,241 @@
...
 # Functions returning a list of parameter arrays
-INITALL_KWARGS_SHAPES = [(qml.init.cvqnn_layers_all, rpt,
-                          [(repeat, n_if), (repeat, n_if), (repeat, n_wires), (repeat, n_wires), (repeat, n_wires),
-                           (repeat, n_if), (repeat, n_if), (repeat, n_wires), (repeat, n_wires), (repeat, n_wires),
-                           (repeat, n_wires)]),
-                         (qml.init.interferometer_all, base, [(n_if,), (n_if,), (n_wires,)])]
+INITALL_KWARGS_SHAPES = [(qml.init.cvqnn_layers_all, {'n_layers': 2, 'n_wires': 3},
+                          [(2, 3), (2, 3), (2, 3), (2, 3), (2, 3),
+                           (2, 3), (2, 3), (2, 3), (2, 3), (2, 3),
+                           (2, 3)]),
```
Comment (antalszava) [inline tests/test_init.py:255]: It will be a reference to the same tuple 11 times, so let me know if that would cause a problem
```suggestion
                          [(2, 3)]*11),
```

File: tests/test_init.py [hunk 6]
```diff
@@ -134,6 +286,11 @@ def test_all_shape(self, init, sgntr, shp, seed):
     def test_same_output_for_same_seed(self, init, sgntr, seed, tol):
         """Confirm that initialization functions return a deterministic output
         for a fixed seed."""
+
+        # exclude case of empty parameter list
+        if len(init(**sgntr).flatten()) == 0:
```
Comment (antalszava) [inline tests/test_init.py:291]: Wait, aren't these the cases when an empty array is to be returned? Could check for that here :)

File: tests/test_init.py (hunk 1 above)
Comment (mariaschuld) [reply tests/test_init.py:28]: You are right, but since this test file is likely to disappear in future, I want to avoid doing any cosmetic improvements at this stage :) 

The edge case tests however were important for functionality - and unearthed indeed one or two problems.

File: pennylane/init.py (hunk 2 above)
Comment (mariaschuld) [reply pennylane/init.py:169]: I know, you are probably right :) I thought a user may iterate over wires and start from 0, but then the qnode would complain...

Maybe it would be better in future to throw an error - will put this on the todo list for the templates upgrade!

File: tests/test_init.py (hunk 3 above)
Comment (mariaschuld) [reply tests/test_init.py:40]: None, only so the user has one for every parameter of the template, without having to know that they are the same.

**Comment by mariaschuld** [comment]
Thanks for review @antalszava .

As mentioned, this test file will disappear within the next few months when we introduce a better strategy for the init functions. The way it is now is a disgrace indeed.   

The PR is more of a "hotfix".

File: tests/test_init.py (hunk 3 above)
Comment (antalszava) [reply tests/test_init.py:40]: I see! Could it make sense to merge them into one instead and the user could e.g. pass a `parameter` argument for the name?

File: tests/test_init.py (hunk 1 above)
Comment (antalszava) [reply tests/test_init.py:28]: Sure!

File: tests/test_init.py (hunk 3 above)
Comment (mariaschuld) [reply tests/test_init.py:40]: Yes, in future this will be the way!
